torch/csrc/dynamo/eval_frame.c

```diff
@@ -848,8 +856,26 @@ inline static PyObject* eval_custom_code(
 
   #endif
 
+  // In CPython 3.11 and above, a custom evaluation frame
+  // means certain optimizations are disabled.
+  // As after this point we are practically evaluating
+  // CPython bytecode using the standard interpreter,
+  // unset the eval frame to let CPython know that it can
+  // perform its optimizations. 
+  #if IS_PYTHON_3_11_PLUS
+  void *prev = tstate->interp->eval_frame;
+  // We are the only one executing this.
+  if (active_dynamo_threads <= 1) {
+    tstate->interp->eval_frame = NULL;
+  }
+  #endif
+
   PyObject* result = eval_frame_default(tstate, shadow, throw_flag);
```

Comment:
This function calls `_PyEval_EvalFrameDefault` directly. I think you need to update it to call `PyEval_EvalFrameEx`? You should probably move the `tstate` manipulation code in there as well.

Reply:
And I'm guessing that respecting `previous_eval_frame` is what's needed rather than the `active_dynamo_threads` check above.

Reply:
I tried that and tests still fail. I have determined that dynamo needs to trace during graph breaks, as what Jason said, so this is not easily feasible https://github.com/pytorch/pytorch/issues/108074#issuecomment-1703821716.